Bluetooth: controller: fix failing EBQ scanner tests

kruithofa opened the pull request:
In the case that chaining is enabled when adding new data to an advertising chain a new PDU is created for this data. This limits how often data can be added to the number of available PDUs.

The changes in this PR add the data to the last PDU in the chain, and only if there is not enough room a new PDU is created

A test was added to CI where the name is set to an extra long length to force an overflow of data to occur so that this new functionality is tested

See issues #63288 and #63290 for further improvements

fixes #59580

Commit:
Bluetooth: controller: fix failing EBQ advertising tests

Updates the chaining for advertising.
Instead of unconditionally adding a new PDU when new data
is added we now instead fill the last PDU in the chain with
the incoming data, only adding a new PDU when there is
not enough room.

This reduces the nr. of PDUs used for advertising, and also
fixes some qualification failures

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Test: Bluetooth: controller: add CI testing for advertising chaining

Update babblesim tests for testing chaining for advertising to ensure that
we spill over to the next PDU

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * the exact amount that would fit
+			 */
+			ad_len_chain = len-ad_len;
+			ad_len_overflow = 0;
+
+		} else {
+			ad_len_overflow = 0;
```
```suggestion
			ad_len_overflow = 0U;

		} else {
			ad_len_overflow = 0U;
```

kruithofa replied:
fixed

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							   0, hdr_data);
+			LL_ASSERT(chain_err == 0);
```
```suggestion
			LL_ASSERT(chain_err == 0U);
```

kruithofa replied:
fixed

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -555,8 +612,12 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
 		/* Fill AD Data in chain PDU */
-		(void)memcpy(dptr_chain, data, len);
-
+		if (ad_len_overflow != 0) {
+			(void)memcpy(dptr_chain, data, ad_len_overflow);
+		}
+		if (ad_len_chain != 0) {
```
```suggestion
		if (ad_len_overflow != 0U) {
			(void)memcpy(dptr_chain, data, ad_len_overflow);
		}
		if (ad_len_chain != 0U) {
```

kruithofa replied:
fixed

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -555,8 +612,12 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			(void)memcpy(dptr_chain, data, ad_len_overflow);
+		}
+		if (ad_len_chain != 0) {
+			(void)memcpy(dptr_chain, &data[ad_len + ad_len_overflow], len);
```
Should this have been
```suggestion
			(void)memcpy(dptr_chain, &data[ad_len + ad_len_overflow], ad_len_chain);
```

kruithofa replied:
fixed

Thalley commented on tests/bsim/bluetooth/host/adv/chain/prj_extra_chain.conf:
```diff
@@ -0,0 +1,41 @@
+CONFIG_BT_EXT_ADV_MAX_ADV_SET=2
+#
+# The extra long device name requires chaining an additional PDU in this test-case
+CONFIG_BT_DEVICE_NAME="Broadcaster Multiple with an extra long name 12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890"
```
Wouldn't it have been "easier" to define some long advertising data in the test instead? The host may truncate this

Thalley replied:
My point is that the name has specific handling in the host: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/adv.c#L511-L518

Where it may be truncated, so using the name to set large advertising data may not work as intended. 

Any other advertising type would work ;)

kruithofa replied:
Removed the long name, and duplicated the manufacturer data field

Thalley commented on tests/bsim/bluetooth/host/adv/chain/prj_extra_chain.conf:
```diff
@@ -0,0 +1,41 @@
+# Set maximum scan data length for Extended Scanning in Bluetooth LE Controller
+CONFIG_BT_CTLR_SCAN_DATA_LEN_MAX=1650
+
+# Zephyr Bluetooth LE Controller needs 16 event buffers to generate Extended
+# Advertising Report for receiving the complete 1650 bytes of data
+CONFIG_BT_BUF_EVT_RX_COUNT=16
```
Why 16? Each report can hold up to 229 octets, so for 1650 octets at most 8 events are required

kruithofa replied:
Hm, this was copied from prj.conf

cvinayak replied:
8 + 8, is used as Zephyr Controller does not combine all the 1650 bytes before fragmenting them into 8 HCI reports, if a PDU has 255 bytes, it will generate 2 HCI reports.

Thalley replied:
I see. Can we update the comment to state this? It just says that it needs 16, but not _why_

kruithofa replied:
Updated the comment to explain why

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -31,7 +31,11 @@
 		bs_trace_info_time(1, __VA_ARGS__);	\
 	} while (0)
 
+#if defined(CONFIG_BT_DEVICE_NAME)
+#define NAME_LEN (strlen(CONFIG_BT_DEVICE_NAME) + 1U)
+#else
 #define NAME_LEN 30
+#endif
```
Ditto

Thalley replied:
@kruithofa Will you update this as well?

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -94,25 +98,29 @@ static void scan_recv(const struct bt_le_scan_recv_info *info,
 
 	data_len = buf->len;
 	if (data_len != DATA_LEN) {
+		printk("Unexpected length\n");
```
Shouldn't this fail? Please also print the `data_len` here for debugging purposes

kruithofa replied:
Didn't cause fail in the original test. Text updated to better reflect this, and print more debugging info

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -94,25 +98,29 @@ static void scan_recv(const struct bt_le_scan_recv_info *info,
 	bt_data_parse(buf, data_cb, name);
 
 	if (strcmp(name, CONFIG_BT_DEVICE_NAME)) {
+		printk("Wrong name\n");
 		return;
 	}
 
 	for (uint8_t i = 0; i < sid_count; i++) {
 		if (sid[i] == info->sid) {
+			printk("Wrong SID\n");
 			return;
 		}
 	}
 
 	sid[sid_count++] = info->sid;
 
 	if (sid_count < CONFIG_BT_EXT_ADV_MAX_ADV_SET) {
+		printk("Wrong number of advertising sets\n");
```
Ditto

kruithofa replied:
Did not fail in the original test. Updated text to reflect this better

kruithofa commented on samples/bluetooth/observer/src/observer.c:
```diff
@@ -9,7 +9,11 @@
 #include <zephyr/bluetooth/bluetooth.h>
 #include <zephyr/bluetooth/hci.h>
 
+#if defined(CONFIG_BT_DEVICE_NAME)
+#define NAME_LEN (strlen(CONFIG_BT_DEVICE_NAME) + 1U)
+#else
 #define NAME_LEN 30
+#endif
```
You are right, we need instead to check whether the device name is configurable during runtime. Updated this piece of code accordingly

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -116,7 +116,9 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
 	struct pdu_adv *pdu_prev;
 	uint8_t ad_len_overflow;
 	uint8_t ad_len_chain;
+	uint8_t chain_err = 0;
```
done

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -317,28 +318,22 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
-				     0U, ULL_ADV_HDR_DATA_DATA_PTR_SIZE);
+			val_ptr = hdr_data;
+			*val_ptr++ = 0U;
+			(void)memset((void *)val_ptr, 0U,
+			     ULL_ADV_HDR_DATA_DATA_PTR_SIZE);
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -317,28 +318,22 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
-				     0U, ULL_ADV_HDR_DATA_DATA_PTR_SIZE);
+			val_ptr = hdr_data;
+			*val_ptr++ = 0U;
+			(void)memset((void *)val_ptr, 0U,
```
OTOH val_ptr points to unsigned int. There are several other places in the code where 0U is used in memset. Let us discuss outside of this PR, and create another PR to fix all locations if that is what we agree on.

Thalley replied:
Doesn't really matter that much :) But it's just common to use `U` when doing unsigned, and omit it for signed values (like for `memset`).

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -317,28 +318,22 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
-						0, hdr_data);
-			LL_ASSERT(!err || (err == BT_HCI_ERR_PACKET_TOO_LONG));
+							ULL_ADV_PDU_HDR_FIELD_AD_DATA,
+							0, hdr_data);
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+		 */
+
+		val_ptr = hdr_data;
+		*val_ptr++ =  len;
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+		/* Append data to the last PDU */
+		chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+						      (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND),
+						      0, hdr_data);
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+
+		/* Append data to the last PDU */
+		chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+						      (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND),
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							    0, hdr_data);
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			val_ptr += sizeof(data);
+			*val_ptr++ =  len;
+			(void)memcpy(val_ptr, &data, sizeof(data));
+			chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							    0, hdr_data);
```
but proper indentation pushes us over the 100 character line limit...

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+							     ULL_ADV_HDR_DATA_DATA_PTR_SIZE];
+
+			val_ptr = hdr_data;
+			*val_ptr++ =  len;
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			*val_ptr++ =  len;
+			(void)memcpy(val_ptr, &data, sizeof(data));
+			val_ptr += sizeof(data);
+			*val_ptr++ =  len;
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							    0, hdr_data);
+			ad_len_overflow =
+				hdr_data[ULL_ADV_HDR_DATA_AUX_PTR_PTR_OFFSET +
+					 ULL_ADV_HDR_DATA_AUX_PTR_PTR_SIZE +
+					 ULL_ADV_HDR_DATA_DATA_PTR_OFFSET +
+					 ULL_ADV_HDR_DATA_DATA_PTR_SIZE];
```
fixed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * is PDU_AC_PAYLOAD_SIZE_MAX wrong?
+			 */
+			ad_len =  PDU_AC_PAYLOAD_SIZE_MAX - ad_len_prev -
+				(ad_len_overflow - ad_len_overflow_first_try) - 4;
```
yes, this needs an explanation @cvinayak

Thalley replied:
@kruithofa If this is correct, please update the comment. 

Regardless if this is correct or not, it probably shouldn't use magic numbers like `4` :)

kruithofa replied:
@cvinayak can we discuss this?

cvinayak replied:
@Tronil and @Emil-Gydesen-Bose  Myself and @kruithofa has a code walk through today and found a bug in the `ull_adv_aux_pdu_set_clear()` function wherein it did not correctly return the overflow length. We now are able to fill the Adv chain PDUs to its max 255 bytes PDU length with @Tronil  suggested `len - ad_len_overflow` as the last PDU's AD length before adding a following new chain PDU to contain the remainder bytes.

Please review #63298, and then this PR will be updated to remove the unnecessary `ad_len_overflow_first_try` value that was used to do the complex calculations.

@kruithofa will update this PR once #63298 is merged.

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,89 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * last PDU without getting an overflow
+			 */
+			chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							   0, hdr_data);
```
either that or over the 100 character limit

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -75,7 +79,7 @@ int broadcaster_multiple(void)
 		return err;
 	}
-
+printk("Length: %d\n", BT_DEVICE_NAME_LEN);
```
Leftover debug?

kruithofa replied:
removed

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,88 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			val_ptr += sizeof(data);
+			*val_ptr++ = len;
+			(void)memcpy(val_ptr, &data, sizeof(data));
+			chain_err = ull_adv_aux_pdu_set_clear(adv, pdu_prev, pdu,
+							   (ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+							    ULL_ADV_PDU_HDR_FIELD_AUX_PTR),
+							    0U, hdr_data);
```
One alternative to make this fit in 100 line length is to do
```suggestion
			chain_err = ull_adv_aux_pdu_set_clear(
				adv, pdu_prev, pdu, 
				(ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND | 
				 ULL_ADV_PDU_HDR_FIELD_AUX_PTR), 
				0U, hdr_data);1
```

Alternatively move `(ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND | ULL_ADV_PDU_HDR_FIELD_AUX_PTR)` to a local `const `variable

Another alternative is to move this into a new function (`ll_adv_aux_ad_data_set` is 500 lines long with this change, so arguably that should be split into multiple functions anyhow).

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,88 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+
+		LL_ASSERT((!chain_err)  || (chain_err == BT_HCI_ERR_PACKET_TOO_LONG));
+
+		/* FIXME: this probably needs a for loop  */
```
updated the comment, the handling has become quite complex and we should probably rewrite this whole procedure to simplify.

kruithofa commented:
@Tronil for some reason I can not reply to your comment regarding the calculation. Actually it is the right calculation, we substract ad_len_overflow_first_try from ad_len_overflow_first to find how much extra data is needed in the variable sized header.
But the code has become quite complicated and @cvinayak and I agree that this code needs a rewrite

Thalley commented:
@kruithofa Please also update the commit message for the first commit (and the PR title too) as "Bluetooth: controller: fix regression" is extremely generic :D

cvinayak reviewed (changes requested):
review rework in ull_adv_aux.c file may have been commited to the test file commits, please move the changes to the other commit.

cvinayak commented:
@aescolar new bsim test, you may be interested to test/review.

aescolar commented:
> @aescolar new bsim test, you may be interested to test/review.

No evident issues to my eyes, and the test works fine on a simulated nrf53

carlescufi commented:
@Thalley PTAL

Thalley commented:
> @Thalley PTAL

@kruithofa There are still a few unresolved comments from previous reviews. Will re-review once they have been resolved

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -75,7 +75,6 @@ int broadcaster_multiple(void)
 		printk("Bluetooth init failed (err %d)\n", err);
 		return err;
 	}
-
```
You can remove this change

Thalley commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -27,7 +27,11 @@
 
 /* Device name length, size minus one null character */
-#define BT_DEVICE_NAME_LEN (sizeof(CONFIG_BT_DEVICE_NAME) - 1U)
+#if defined(CONFIG_BT_DEVICE_NAME_DYNAMIC)
+#define BT_DEVICE_NAME_LEN (CONFIG_BT_DEVICE_NAME_MAX + 1U)
+#else
+#define BT_DEVICE_NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
+#endif
```
This change now seems unnecessary

kruithofa replied:
It is not really necessary, but is a more generic way of setting the device name. I prefer leaving it this way.

Thalley commented on samples/bluetooth/observer/src/observer.c:
```diff
@@ -9,7 +9,11 @@
 #include <zephyr/bluetooth/hci.h>
 
-#define NAME_LEN 30
+#if defined(CONFIG_BT_DEVICE_NAME_DYNAMIC)
+#define NAME_LEN (CONFIG_BT_DEVICE_NAME_MAX + 1U)
+#else
+#define NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
+#endif
```
Ditto; this change can be reverted

kruithofa replied:
The original set the NAME_LEN to a fixed nr. I prefer this way

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -31,11 +31,19 @@
 	} while (0)
 
-#define NAME_LEN 30
+#if defined(CONFIG_BT_DEVICE_NAME_DYNAMIC)
+#define NAME_LEN (CONFIG_BT_DEVICE_NAME_MAX + 1U)
+#else
+#define NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
+#endif
```
Ditto; this change can be reverted

kruithofa replied:
Ditto I will remove the conditional compile

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,88 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * ad_len_overflow is how much overflows with the AUX ptr
+			 */
+			uint8_t ad_len_overflow_first_try;
+			const uint16_t chain_add_fields = ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
+				ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
```
```suggestion
			const uint16_t chain_add_fields = ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND |
							  ULL_ADV_PDU_HDR_FIELD_AUX_PTR;
```

kruithofa commented:
we want to consider issue #63290 and #63288 before merging this (or replacing with another PR)

Commit:
Bluetooth: controller: change calc for data in PDU

The existing formula for the amount of data that can be filled
in in the current PDU is correct but confusing. Replace it with
a simpler formula.

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Thalley commented on samples/bluetooth/observer/src/observer.c:
```diff
@@ -9,7 +9,7 @@
 #include <zephyr/bluetooth/hci.h>
 
-#define NAME_LEN 30
+#define NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
```
Can't this be reverted back to how it was? Or is this change necessary?

kruithofa replied:
Ditto, separate PR

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,85 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+						      ULL_ADV_PDU_HDR_FIELD_AD_DATA_APPEND,
+						      0U, hdr_data);
+
+		LL_ASSERT((!chain_err)  || (chain_err == BT_HCI_ERR_PACKET_TOO_LONG));
```
```suggestion
		LL_ASSERT((!chain_err) || (chain_err == BT_HCI_ERR_PACKET_TOO_LONG));
```

kruithofa replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,85 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * FIXME: ad_len_overflow already should be 0, to be verified. We wait
+			 * fixing this until rewriting this whole function
+			 */
+			ad_len_overflow = 0U;
+
```
```suggestion
			ad_len_overflow = 0U;
```

kruithofa replied:
done

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -555,8 +608,12 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
-
+		if (ad_len_overflow != 0U) {
+			(void)memcpy(dptr_chain, data, ad_len_overflow);
+		}
+		if (ad_len_chain != 0U) {
```
```suggestion
		}
		
		if (ad_len_chain != 0U) {
```

kruithofa replied:
done

Thalley commented on tests/bsim/bluetooth/host/adv/chain/src/main.c:
```diff
@@ -31,11 +31,15 @@
 	} while (0)
 
-#define NAME_LEN 30
+#define NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
```
Can't this be reverted back to how it was? Or is this change necessary?

kruithofa replied:
Ditto, separate PR

kruithofa commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -26,8 +26,8 @@
 
-/* Device name length, size minus one null character */
-#define BT_DEVICE_NAME_LEN (sizeof(CONFIG_BT_DEVICE_NAME) - 1U)
+/* Device name length, size */
+#define BT_DEVICE_NAME_LEN sizeof(CONFIG_BT_DEVICE_NAME)
```
it is not related to what we are fixing here, but is more correct. I'll put it in a separate PR

Tronil commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,84 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+		if (chain_err == BT_HCI_ERR_PACKET_TOO_LONG) {
+			/* We could not fit all the data, append as much as possible
+			 * ad_len_overflow_first_try is how much overflows without the AUX ptr
+			 * ad_len_overflow is how much overflows with the AUX ptr
```
Unless I am missing something, `ad_len_overflow_first_try` was removed, so this comment is no longer correct

kruithofa replied:
fixed

Tronil commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -411,8 +385,84 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
+			 * the difference of the two is the amount that
+			 * we can fit in the current PDU
+			 */
+			ad_len = len - ad_len_chain;
```
What happens if we have a completely full PDU that we want to add an aux ptr to (ie. we now have to move some existing advertising data to the next PDU)? This code doesn't seem to handle that case.

kruithofa replied:
That is handled by the ull_adv_aux_clear_set function. However this code is hard to follow, so we want to do a rewrite. See issue #63288 and #63290

cvinayak commented on samples/bluetooth/broadcaster_multiple/src/broadcaster_multiple.c:
```diff
@@ -47,11 +47,18 @@
-
+/*
+ * Datalength is an integer, so BT_MFG_DATA_LEN can not be larger than 255.
+ * To ensure that we need to chain PDUs we therefor add manufacturer data
```
```suggestion
 * To ensure that we need to chain PDUs we therefore add manufacturer data
```

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -116,7 +116,9 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
 	struct pdu_adv *pdu_prev;
 	uint8_t ad_len_overflow;
 	uint8_t ad_len_chain;
+	uint8_t chain_err = 0U;
```
Move this declaration as close to its locality of reference.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_adv_aux.c:
```diff
@@ -116,7 +116,9 @@ uint8_t ll_adv_aux_ad_data_set(uint8_t handle, uint8_t op, uint8_t frag_pref,
 	uint8_t ad_len_chain;
+	uint8_t chain_err = 0U;
 	struct pdu_adv *pdu;
+	uint8_t ad_len = 0U, ad_len_prev = 0U;
```
May be these too can move closer to locality of reference? (to reduce program stack depth)

kruithofa replied:
only ad_len_prev can be moved
